LogicLayer/Friends/FriendsService.cs

```diff
@@ -0,0 +1,133 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data.SqlClient;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading.Tasks;
+using GameBoard.DataLayer.Entities;
+using GameBoard.DataLayer.Enums;
+using GameBoard.LogicLayer.Friends.Dtos;
+using GameBoard.LogicLayer.UserSearch.Dtos;
+using GameBoard.DataLayer.Repositories;
+using GameBoard.LogicLayer.Friends.Exceptions;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Query.Expressions;
+using Microsoft.Extensions.Logging;
+using Remotion.Linq.Parsing;
+
+namespace GameBoard.LogicLayer.Friends
+{
+    internal sealed class FriendsService : IFriendsService
+    {
+        private readonly IGameBoardRepository _repository;
+
+        public FriendsService(IGameBoardRepository repository)
+        {
+            _repository = repository;
+        }
+
+        public async Task<IEnumerable<UserDto>> GetFriendsByUserNameAsync(string userName)
+        {
+            string normalizedUserName = userName.ToUpper();
+            var user = _repository.ApplicationUsers.Where(u => u.NormalizedUserName == normalizedUserName);
+
+            var friendsInvitedByMe = user
+                .Include(u => u.SentRequests)
+                .SelectMany(u => u.SentRequests)
+                .Where(f => f.FriendshipStatus == FriendshipStatus.Lasts)
+                .Include(f => f.RequestedTo)
+                .Select(f => new UserDto(f.RequestedTo.Id, f.RequestedTo.UserName, f.RequestedTo.Email));
+
+            var friendsThatInvitedMe = user
+                .Include(u => u.ReceivedRequests)
+                .SelectMany(u => u.ReceivedRequests)
+                .Where(f => f.FriendshipStatus == FriendshipStatus.Lasts)
+                .Include(f => f.RequestedBy)
+                .Select(f => new UserDto(f.RequestedBy.Id, f.RequestedBy.UserName, f.RequestedBy.Email));
+
+            var allFriends = await friendsInvitedByMe // can I avoid using await here?
+                .Union(friendsThatInvitedMe)
+                .OrderBy(u => u.UserName) // maybe order by NormalizedUserName? However, I cannot do that with UserDto.
+                .ToListAsync();
+
+            return allFriends;
+        }
+
+        public async Task SendFriendRequestAsync(CreateFriendRequestDto friendRequest)
+        {
+            if (friendRequest.UserNameTo == friendRequest.UserNameFrom)
+            {
+                throw new ApplicationException("You cannot invite yourself.");
```

Comment:
User code should never throw generic `ApplicationExceptions`. I created two exceptions in the `LogicLayer.Exceptions` already, but in hindsight I think it would be best to have a single `InvalidFriendRequestException` that represents a failed attempt at adding a friend request and give them meaningful exception messages, but you can go along and add more detailed exceptions. Anyway - not a generic `ApplicationException`, derive from it.

Reply:
Did it.